core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -179,31 +179,31 @@ class GroupCoordinator(val brokerId: Int,
 
         if (group.hasStaticMember(groupInstanceId)) {
           val oldMemberId = group.getStaticMemberId(groupInstanceId)
+          info(s"Static member $groupInstanceId with unknown member id rejoins, assigning new member id $newMemberId, while " +
+            s"old member $oldMemberId will be removed. It group is currently stable, no rebalance will be triggered.")
 
-          if (group.is(Stable)) {
-            info(s"Static member $groupInstanceId with unknown member id rejoins, assigning new member id $newMemberId, while " +
-              s"old member $oldMemberId will be removed. No rebalance will be triggered.")
+          val oldMember = group.replaceGroupInstance(oldMemberId, newMemberId, groupInstanceId)
+          // Heartbeat of old member id will expire without affection since the group no longer contains that member id.
+          // New heartbeat shall be scheduled with new member id.
+          completeAndScheduleNextHeartbeatExpiration(group, oldMember)
 
-            val oldMember = group.replaceGroupInstance(oldMemberId, newMemberId, groupInstanceId)
+          val knownStaticMember = group.get(newMemberId)
+          group.updateMember(knownStaticMember, protocols, responseCallback)
 
-            // Heartbeat of old member id will expire without affection since the group no longer contains that member id.
-            // New heartbeat shall be scheduled with new member id.
-            completeAndScheduleNextHeartbeatExpiration(group, oldMember)
-
-            responseCallback(JoinGroupResult(
-              members = if (group.isLeader(newMemberId)) {
-                group.currentMemberMetadata
-              } else {
-                List.empty
-              },
-              memberId = newMemberId,
-              generationId = group.generationId,
-              subProtocol = group.protocolOrNull,
-              leaderId = group.leaderOrNull,
-              error = Errors.NONE))
-          } else {
-            val knownStaticMember = group.get(oldMemberId)
-            updateMemberAndRebalance(group, knownStaticMember, protocols, responseCallback)
+          group.currentState match {
+            case Stable =>
```

Comment:
Should this be `Stable | CompletingRebalance`? The join phase of a rebalance may be completed.

Reply:
For `CompletingRebalance` then we should do more: we will respond normally to this member, while for the old member.id, if there's a responseCallback already registered, we should return that response with FENCED error immediately. Otherwise that old member's request would eventually timed out and retry which is unnecessary overhead.

Reply:
There shouldn't be any member with pending join response at `CompletingRebalance` stage IIUC. But the fencing idea is great and I already added it to the `replaceGroupInstance`